Update 98_exercises.qmd (Changes to #512)

jkc9886 opened the pull request:
This PR is to accommodate the changes requested in PR#512 which was closed recently. Changes made:

1. Removed requested text(line 123 and 1215).
2. Edited the Heatmap annotation exercise by using draw function from sechm package to create heatmap directly from tse.

jkc9886 committed:
Update 98_exercises.qmd (Changes to #512)

This PR is to accommodate the changes requested in #512 :

1. Removed the requested text (Line 123 and 1215).
2. Edited the Heatmap annotation exercise by using draw function from sechm package to create heatmap directly from tse.

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1190,31 +1187,31 @@ PCoA object).
 # Load necessary libraries
-library(SummarizedExperiment)
-library(pheatmap)
+library(TreeSummarizedExperiment)
```
Is mia already loaded (I believe it is)? It loads automatically TreeSE so this line can be removed

TuomasBorman replied:
Well since this is a solution for example, it is good to leave here. But load mia package instead of TreeSummarizedExperiment

jkc9886 replied:
Sure, noted :))

antagomir replied:
@jkc9886 have you fixed this now? if yes, kindly resolve the comment

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1190,31 +1187,31 @@ PCoA object).
-pheatmap(assay_data, cluster_rows = TRUE, cluster_cols = TRUE, annotation_col = as.data.frame(sample_metadata))
+row_data <- DataFrame(GeneID = 1:20) # Row metadata
+phylo_tree <- rtree(n=20) # Generating a random phylogenetic tree for demonstration
+
+# Extract sample metadata for annotation directly from the tse object
+sample_metadata <- colData(tse)   
+
+# Create TreeSummarizedExperiment object
+tse <- TreeSummarizedExperiment(assays = list(counts = example_data),
+                                       colData = col_data,
+                                       rowData = row_data,
+                                       rowTree = phylo_tree)
```
This is rather complex and is not central for this exercise. So I would stick with datasets found from mia

TuomasBorman replied:
Use the same dataset that is used above (pcoa_bray). However, the name of that variable is not good. It should be tse.

TuomasBorman replied:
The idea of TreeSummarizedExperiment is to store the results to same object and not to create new pbject for every result

TuomasBorman reviewed (changes requested):
See comments

jkc9886 committed:
Update 98_exercises.qmd

1. Exchanged 2 heatmap visualization exercises ( clustering and annotations) with 3 exercises ( hierarchial clustering, k-means clustering and annotations) to make it clearer to understand. 

 ( clustering and annotations) ->  ( hierarchial clustering, k-means clustering and annotations)

@TuomasBorman is the naming consistent as suggested by you in PR #518 

2. Question : Should the diagrams of these heatmaps (outputs) be included in OMA or is the current version(without the diagram) fine? Please build it locally once and guide me on what else could be done.

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+library(SummarizedExperiment)
+library(vegan)
+library(ggplot2)
+library(cluster)
```
Remove cluster and ggplot2 packages

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+library(cluster)
+
+# Generate a sample SummarizedExperiment object
+data(GlobalPatterns)
+se <- SummarizedExperiment(assays = list(counts = assay(GlobalPatterns)),
+                           colData = colData(GlobalPatterns),
+                           rowData = rowData(GlobalPatterns))
```
You can just use the object as it is

tse <- GlobalPatterns

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+                           colData = colData(GlobalPatterns),
+                           rowData = rowData(GlobalPatterns))
+
+# Perform PCoA on the abundance data
+distance_matrix <- vegdist(t(assay(se)), method = "bray")
+pcoa_results <- cmdscale(distance_matrix, k = 2)
+pcoa_data <- data.frame(pcoa_results)
+colnames(pcoa_data) <- c("PCoA1", "PCoA2")
```
Instead, use runMDS

jkc9886 replied:
I am facing a continuous error which indicates that the runMDS function is not available in the mia package, could you help me with me with it?

TuomasBorman replied:
Try scater::runMDS()

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+pcoa_data <- data.frame(pcoa_results)
+colnames(pcoa_data) <- c("PCoA1", "PCoA2")
+
+# Perform k-means clustering (k = 3)
+set.seed(123)
+kmeans_result <- kmeans(pcoa_data, centers = 3)
+pcoa_data$cluster <- as.factor(kmeans_result$cluster)
```
Instead use bluster package

jkc9886 replied:
Changes made in the latest commit : 

1. In heatmap visualization ex-1:
- `scater::runMDS` is used to perform MDS on the abundance data.
- `bluster` package is used for hierarchical clustering and k-means clustering.
- `ComplexHeatmap` < – > “pheatmap”

2. In heatmap visualization ex-2:
- Changed to CLR+Euclidean 

3. In heatmap visualization ex-3:
- Assumes that the cluster information has already been added to `colData` and does not recalculate distances and clusters.

Just a question, Is there a specific reason we are using complexheatmap

TuomasBorman replied:
pheatmap and ComplexHeatmap are convenient tools to plot heatmaps. They are easier to use than ggplot2 (for making nicer heatmaps). ComplexHeatmap is the next iteration of pheatmap, it can do the same as pheatmap and more. In fact, it has ComplexHeatmap::pheatmap function that creates pheatmap-like plots.

sechm is a package to plot assay of SummarizedExperiment. It creates ComplexHeatmap object. We should use sechm whenever possible since it makes the examples simpler (user can always use ComplexHeatmap directly if wants).

jkc9886 replied:
Okay I got what you mean now, thank you :)

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+kmeans_result <- kmeans(pcoa_data, centers = 3)
+pcoa_data$cluster <- as.factor(kmeans_result$cluster)
+
+# Visualize PCoA with clusters
+ggplot(pcoa_data, aes(x = PCoA1, y = PCoA2, color = cluster)) +
+  geom_point(size = 3) +
+  labs(title = "PCoA with k-means Clustering", x = "PCoA Axis 1", y = "PCoA Axis 2") +
+  theme_minimal() +
+  scale_color_manual(values = c("red", "blue", "green"))
```
Instead use plotReducedDim

TuomasBorman commented on inst/pages/98_exercises.qmd:
````diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
 ```{r}
+#| label: heatmap-visualization-ex1
+#| eval: FALSE
+#| code-fold: true
+#| code-summary: "Show solution"
+
 # Load necessary libraries
+library(mia)
 library(SummarizedExperiment)
 library(pheatmap)
+library(cluster)
+
+# Generate a sample SummarizedExperiment object 
+data(GlobalPatterns)
+se <- SummarizedExperiment(assays = list(counts = assay(GlobalPatterns)),
+                           colData = colData(GlobalPatterns),
+                           rowData = rowData(GlobalPatterns))
+
+# Extract the abundance data matrix
+abundance_data <- assay(se, "counts")
+
+# Perform hierarchical clustering on the samples
+distance_matrix <- dist(t(abundance_data), method = "euclidean")
+clustering <- hclust(distance_matrix, method = "average")
+
+# Cut the dendrogram to form clusters (k = 3 for this example)
+clusters <- cutree(clustering, k = 3)
+
+# Add cluster information to colData
+colData(se)$cluster <- as.factor(clusters)
+
+# Visualize the heatmap with clusters
+heatmap_data <- assay(se, "counts")
+pheatmap(heatmap_data, 
+         cluster_rows = clustering, 
+         cluster_cols = clustering, 
+         annotation_col = data.frame(Cluster = colData(se)$cluster),
+         main = "Heatmap with Clustering",
+         show_rownames = FALSE, 
+         show_colnames = FALSE)
````
Same here. Also use ComplexHeatmap instead of pheatmap

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+library(SummarizedExperiment)
+library(pheatmap)
 
+# Generate a sample SummarizedExperiment object
+data(GlobalPatterns)
+se <- SummarizedExperiment(assays = list(counts = assay(GlobalPatterns)),
+                           colData = colData(GlobalPatterns),
+                           rowData = rowData(GlobalPatterns))
+
+# Extract the abundance data matrix
+abundance_data <- assay(se, "counts")
+
+# Ensure cluster information is correctly added to colData
+# Perform hierarchical clustering on the samples
+distance_matrix <- dist(t(abundance_data), method = "euclidean")
+clustering <- hclust(distance_matrix, method = "average")
+
+# Cut the dendrogram to form clusters (k = 3 for this example)
+clusters <- cutree(clustering, k = 3)
+
+# Add cluster information to colData
+colData(se)$cluster <- as.factor(clusters)
+
+# Create a sample metadata annotation (e.g., SampleType)
+sample_metadata <- data.frame(SampleType = colData(se)$SampleType,
+                              Cluster = colData(se)$cluster)
+rownames(sample_metadata) <- colnames(abundance_data)
+
+# Visualize the heatmap with annotations
+pheatmap(abundance_data, 
+         annotation_col = sample_metadata,
+         main = "Heatmap with Annotations",
+         show_rownames = FALSE, 
+         show_colnames = FALSE)
```
No need to calculate everything again

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+                           rowData = rowData(GlobalPatterns))
+
+# Extract the abundance data matrix
+abundance_data <- assay(se, "counts")
+
+# Perform hierarchical clustering on the samples
+distance_matrix <- dist(t(abundance_data), method = "euclidean")
```
Usually this is not done for counts + euclidean. Common choice is CLR + euclidean or relative abundance + Bray-Curtis

jkc9886 replied:
Okay! I am working on the revised code :)

jkc9886 committed:
Update 98_exercises.qmd

Changes made in this commit : 

1. In heatmap visualization ex-1:
- `scater::runMDS` is used to perform MDS on the abundance data.
- `bluster` package is used for hierarchical clustering and k-means clustering.
- `ComplexHeatmap` < – > “pheatmap”

2. In heatmap visualization ex-2:
- Changed to CLR+Euclidean 

3. In heatmap visualization ex-3:
- Assumes that the cluster information has already been added to `colData` and does not recalculate distances and clusters.

Just a question, Is there a specific reason we are using complexheatmap?

jkc9886 commented:
This PR is under progress.

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,150 @@ See the OMA book for examples.
-# Example data creation
-# Assuming enterotype is a SummarizedExperiment object with sample metadata
-# and assay_data is a matrix containing assay data
+# Extract the abundance data matrix
+abundance_data <- assay(se, "counts")
+
+# Perform CLR transformation
+clr_data <- t(apply(abundance_data, 1, function(x) {
+  compositions::clr(x + 1)  # Adding 1 to avoid log(0)
+}))
```
Just use tranformAssay

jkc9886 replied:
Could you please check the latest commit whenever free? I made the change.

jkc9886 committed:
Update 98_exercises.qmd

Line 1226, used transformAssay to directly use clr_data.

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
-# Assuming enterotype is a SummarizedExperiment object with sample metadata
-# and assay_data is a matrix containing assay data
+# Extract the abundance data matrix and perform CLR transformation
+clr_data <- transformAssay(se, method = "clr", assay = "counts")
```
I have not run this line but does this work? clr does require that there are no zeroes in the data, and if I remember correctly GlobalPatterns has them. (use pseudocount)

Also:

- assay --> assay.type
- se --> tse
- replace the old variable with new one

Probably:

tse <- transformAssay(tse, assay.type = "counts", method = "clr", pseudocount = TRUE)

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+clr_data <- transformAssay(se, method = "clr", assay = "counts")
+
+# Perform hierarchical clustering on the samples using Euclidean distance
+distance_matrix <- dist(clr_data, method = "euclidean")
```
clr_data should be TreeSE object as it comes from transformAssay(). stats::dist requires matrix object which indicates that this line is not working.

TuomasBorman replied:
Please run the code and check that it is working

jkc9886 replied:
It works when I build it on my local system.

antagomir replied:
This clearly does not work. 

```
# Get example data
library(mia)
data(GlobalPatterns)
tse <- GlobalPatterns
se <- GlobalPatterns

# Try this, doesn't work:
clr_data <- transformAssay(se, method = "clr", assay = "counts")
#Error in transformAssay(se, method = "clr", assay = "counts") : 
# argument 3 matches multiple formal arguments

# Ok first issue is that there is no argument "assay" in transformAssay function.. so let's change it:
clr_data <- transformAssay(se, method = "clr", assay.type = "counts")
#Error: 'clr' cannot be used with non-positive data: use pseudocount > 0

# Ok pseudocount was also missing, let's add:
clr_data <- transformAssay(se, method = "clr", assay.type = "counts", pseudocount=TRUE)
# A pseudocount of 0.5 was applied.

# This also doesn't work for SE objects
distance_matrix <- dist(clr_data, method = "euclidean")
#Error in h(simpleError(msg, call)) : 
#  error in evaluating the argument 'x' in selecting a method for function 'as.matrix': no method for coercing #this S4 class to a vector

# It does work for matrix (assay) as Tuomas said:
distance_matrix <- dist(assay(clr_data, "clr"), method = "euclidean")
```

-> Run all code in order to test it before pushing to PR

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+colData(se)$cluster <- as.factor(clusters)
+
+# Visualize the heatmap with clusters using ComplexHeatmap
+Heatmap(clr_data, 
+        name = "CLR Counts",
```
Use schem package since you are plotting abundance data

jkc9886 replied:
I rewrote the codes for exercise 2 and 3 using sechm package, could you review it once when possible?

antagomir replied:
Seems ok

TuomasBorman commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
+                                col = list(Cluster = c("1" = "red", "2" = "blue", "3" = "green")))
+
+# Visualize the heatmap with annotations using ComplexHeatmap
+Heatmap(assay(se, "counts"), 
```
Use sechm package

TuomasBorman replied:
https://www.bioconductor.org/packages/release/bioc/html/sechm.html

jkc9886 committed:
Update 98_exercises.qmd

Rewritten the code for heatmap visualization exercise 2 & 3 using sechm package.

antagomir commented:
Can you resolve the git conflicts?

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
 To make the exercise more engaging, let's do an exploratory data analysis task:
 
-#### 1. Cluster Analysis on PCoA/PCA Results:
+### 1: k-means Clustering on Reduced Dimensions and Visualization
```
Please do not capitalize words unnecessarily

antagomir replied:
Can you resolve conversations that have been dealt with?

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
-'plotReducedDim' is a function that can plot these reduced dimensions and has a 'colour_by argument' that accepts a vector of cluster assignments.
 
-#### 2. Heatmap with Annotations:
+### 2: Cluster Analysis with Heatmap Visualization
```
do not capitalize the words

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
+data(GlobalPatterns)
+tse <- GlobalPatterns
+
+# CLR transformation of abundance data matrix
+clr_data <- as.matrix(transformAssay(tse, method = "clr", assay.type = "counts", pseudocount = TRUE))
+
+# Hierarchical clustering on samples
+clustering <- hclust(dist(clr_data, method = "euclidean"), method = "average")
+clusters <- cutree(clustering, k = 3)
+colData(tse)$cluster <- as.factor(clusters)
+
```
Is it necessary to show how to do this with base R as we have the clusterRows (and addCluster() ) function?

jkc9886 replied:
I get what you mean, I thought we should show every basic step since it was a new exercise. I shall make the corrections, thank you :)

antagomir replied:
Yes here we want to teach how to use mia tools so we demonstrate those.

By the way, here line 1217 is not possible: the output from transformAssay is a SE object and I don't think that it accepts as.matrix() command. See for instance:

`as.matrix(GlobalPatterns)`

> Error in h(simpleError(msg, call)) : 
>   error in evaluating the argument 'x' in selecting a method for function 'as.matrix': no method for coercing this S4 class to a vector
> 


-> Did you try the code before pushing the PR?

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
-# Plot the reduced dimensions of pca_aitchison with clusters as colors
-plotReducedDim(pca_aitchison, colour_by = pca_aitchison_clusters)
+# Perform k-means clustering (k = 3) using the bluster package
+set.seed(123)
+clustering_result <- clusterRows(reducedDim(tse, "MDS"), BLUSPARAM=KmeansParam(centers=3))
+colData(tse)$cluster <- as.factor(clustering_result)
```
Could we just use `addCluster()` instead?

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
+  column_labels = FALSE,
+  row_cluster = clustering,
+  column_cluster = clustering,
+  row_annotation = annotation_df,
```
Whynot just `colData(tse)` or `as.data.frame(colData(tse))`?

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
+colData(tse)$cluster <- as.factor(clusters)
+
+# Create annotation for clusters
+annotation_df <- data.frame(Cluster = colData(tse)$cluster)
```
Possibly not needed, see comment on line 1235

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
+
+# Visualize the heatmap with sechm
+sechm(
+  mat = clr_data,
```
whynot just `mat = assay(tse, "clr")`

TuomasBorman replied:
sechm takes SummarizdExperiment as input. Does it even have "mat" parameter?

sechm(tse, assayName = "clr")

TuomasBorman replied:
See

https://www.bioconductor.org/packages/release/bioc/vignettes/sechm/inst/doc/sechm.html

and

?sechm::sechm

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
+  mat = clr_data,
+  row_labels = FALSE,
+  column_labels = FALSE,
+  row_cluster = clustering,
+  column_cluster = clustering,
```
Can we get these directly from the tse object after running `addCluster()`?

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
+# Ensure cluster information is correctly added to colData
+# (Assuming the clusters have been calculated previously and added to colData)
+
+# Create a sample metadata annotation (e.g., SampleType and Cluster)
+sample_metadata <- data.frame(SampleType = colData(tse)$SampleType,
+                              Cluster = colData(tse)$cluster)
+rownames(sample_metadata) <- colnames(assay(tse, "counts"))
```
Is this really needed when we can call `as.data.frame(colData(tse))` or perhaps even just `colData(tse)`

antagomir commented on inst/pages/98_exercises.qmd:
```diff
@@ -1158,62 +1155,128 @@ See the OMA book for examples.
+
+# Visualize the heatmap with annotations using sechm
+sechm_plot <- sechm(assay(tse, "counts"),
+                    annotation_col = sample_metadata,
```
Just do `colData(tse)`

jkc9886 committed:
Update 98_exercises.qmd

1. The clr data is added to the tse object using - assay(tse, "clr") <- as.matrix(clr_data).

2. Clusters added to the tse object using addCluster.

3. The mat parameter in sechm -> assay(tse, "clr").

4. object.annotation_col: Directly uses as.data.frame(colData(tse)) to extract metadata annotations from tse.

antagomir commented:
Resolve the conflicts, too.
